File: src/main/kotlin/models/Ambiente.kt
package models

import org.gustavolyra.portugolpp.PortugolPPParser

class Ambiente(val enclosing: Ambiente? = null) {
    private val valores = mutableMapOf<String, Valor>()

    private val classes = mutableMapOf<String, PortugolPPParser.DeclaracaoClasseContext>()

    private val interfaces = mutableMapOf<String, PortugolPPParser.DeclaracaoInterfaceContext>()

    //referente a instancia que estamos...
    var thisObjeto: Valor.Objeto? = null

    fun definirInterface(nome: String, declaracao: PortugolPPParser.DeclaracaoInterfaceContext) {
        interfaces[nome] = declaracao
    }

    fun obterInterface(nome: String): PortugolPPParser.DeclaracaoInterfaceContext? {
        return interfaces[nome] ?: enclosing?.obterInterface(nome)
    }

    fun getInterfaces(classeContext: PortugolPPParser.DeclaracaoClasseContext): List<String> {
        val result = mutableListOf<String>()
        var foundImplements = false
        for (i in 0 until classeContext.childCount) {
            val token = classeContext.getChild(i).text
            if (foundImplements) {
                if (token == "{") break
                if (token != "," && token != "implementa") {
                    result.add(token)
                }
            } else if (token == "implementa") {
                foundImplements = true
            }
        }
        return result
    }

    fun getSuperClasse(classeContext: PortugolPPParser.DeclaracaoClasseContext): String? {
        for (i in 0 until classeContext.childCount) {
            if (classeContext.getChild(i).text == "estende" && i + 1 < classeContext.childCount) {
                return classeContext.getChild(i + 1).text
            }
        }
        return null
    }


    fun definir(nome: String, valor: Valor) {
        valores[nome] = valor
    }

    fun obter(nome: String): Valor {
        if (nome == "this" && thisObjeto != null) return thisObjeto!!
        val valor = valores[nome]
        if (valor != null) {
            return valor
        }

        if (thisObjeto != null) {
            val campoValor = thisObjeto!!.campos[nome]
            if (campoValor != null) {
                return campoValor
            }
        }

        val externoValor = enclosing?.obter(nome)
        if (externoValor != null && externoValor != Valor.Nulo) {
            return externoValor
        }
        return Valor.Nulo
    }

    fun definirClasse(nome: String, declaracao: PortugolPPParser.DeclaracaoClasseContext) {
        classes[nome] = declaracao
    }

    fun obterClasse(nome: String): PortugolPPParser.DeclaracaoClasseContext? {
        return classes[nome] ?: enclosing?.obterClasse(nome)
    }

    fun atualizarOuDefinir(nome: String, valor: Valor) {
        var ambienteAtual: Ambiente? = this
        while (ambienteAtual != null) {
            if (ambienteAtual.valores.containsKey(nome)) {
                // Encontrou a variável, atualiza no escopo correto
                ambienteAtual.valores[nome] = valor
                return
            }
            ambienteAtual = ambienteAtual.enclosing
        }
        valores[nome] = valor
    }
}

File: cpp_src/src/Ambiente.cpp
#include "Ambiente.h"
#include <stdexcept>

Ambiente::Ambiente(std::shared_ptr<Ambiente> enc) : enclosing(enc), thisObjeto(nullptr) {}

void Ambiente::definir(const std::string& nome, std::shared_ptr<Valor> valor) {
    valores[nome] = valor;
}

std::shared_ptr<Valor> Ambiente::obter(const std::string& nome) {
    if (nome == "this" && thisObjeto != nullptr) {
        return thisObjeto;
    }
    
    auto it = valores.find(nome);
    if (it != valores.end()) {
        return it->second;
    }
    
    // Check if it's a field of the current object
    if (thisObjeto != nullptr) {
        auto& campos = thisObjeto->getCampos();
        auto campoIt = campos.find(nome);
        if (campoIt != campos.end()) {
            return campoIt->second;
        }
    }
    
    // Check enclosing environment
    if (enclosing != nullptr) {
        auto externoValor = enclosing->obter(nome);
        if (externoValor != nullptr && externoValor->getType() != Valor::Type::NULO) {
            return externoValor;
        }
    }
    
    return VALOR_NULO;
}

void Ambiente::atualizarOuDefinir(const std::string& nome, std::shared_ptr<Valor> valor) {
    // Look for the variable in this and enclosing environments
    std::shared_ptr<Ambiente> ambienteAtual = shared_from_this();
    while (ambienteAtual != nullptr) {
        auto it = ambienteAtual->valores.find(nome);
        if (it != ambienteAtual->valores.end()) {
            ambienteAtual->valores[nome] = valor;
            return;
        }
        ambienteAtual = ambienteAtual->enclosing;
    }
    
    // If not found, define in current environment
    valores[nome] = valor;
}

void Ambiente::definirClasse(const std::string& nome, antlr4::tree::ParseTree* declaracao) {
    classes[nome] = declaracao;
}

antlr4::tree::ParseTree* Ambiente::obterClasse(const std::string& nome) {
    auto it = classes.find(nome);
    if (it != classes.end()) {
        return it->second;
    }
    
    if (enclosing != nullptr) {
        return enclosing->obterClasse(nome);
    }
    
    return nullptr;
}

void Ambiente::definirInterface(const std::string& nome, antlr4::tree::ParseTree* declaracao) {
    interfaces[nome] = declaracao;
}

antlr4::tree::ParseTree* Ambiente::obterInterface(const std::string& nome) {
    auto it = interfaces.find(nome);
    if (it != interfaces.end()) {
        return it->second;
    }
    
    if (enclosing != nullptr) {
        return enclosing->obterInterface(nome);
    }
    
    return nullptr;
}

std::vector<std::string> Ambiente::getInterfaces(antlr4::tree::ParseTree* classeContext) {
    // This would need to be implemented with ANTLR parse tree navigation
    // For now, return empty vector
    return {};
}

std::string Ambiente::getSuperClasse(antlr4::tree::ParseTree* classeContext) {
    // This would need to be implemented with ANTLR parse tree navigation
    // For now, return empty string
    return "";
}

void Ambiente::setThisObjeto(std::shared_ptr<ValorObjeto> obj) {
    thisObjeto = obj;
}

std::shared_ptr<ValorObjeto> Ambiente::getThisObjeto() const {
    return thisObjeto;
}

std::shared_ptr<Ambiente> Ambiente::getEnclosing() const {
    return enclosing;
}